src/common/src/types/num256.rs

```diff
@@ -125,15 +164,263 @@ macro_rules! impl_common_for_num256 {
+        deserializer: &mut memcomparable::Deserializer<impl Buf>,
+    ) -> memcomparable::Result<Self> {
+        let buf = deserializer.deserialize_bytes(FormattedInt256Visitor)?;
+        let unsigned = i256::from_be_bytes(buf);
+        let signed = unsigned ^ (i256::from(1) << (i256::BITS - 1));
+        Ok(Int256::from(signed))
+    }
+}
+
+macro_rules! impl_convert_from {
+    ($($t:ty),* $(,)?) => {$(
+        impl From<$t> for Int256 {
+            #[inline]
+            fn from(value: $t) -> Self {
+                Self(Box::new(i256::from(value)))
+            }
+        }
+    )*};
+}
+
+impl_convert_from!(i16, i32, i64);
+
+// Conversions for mathematical operations.
+impl From<Int256Ref<'_>> for Int256 {
+    fn from(value: Int256Ref<'_>) -> Self {
+        value.to_owned_scalar()
+    }
+}
+
+macro_rules! impl_checked_op {
+    ($trait:ty, $func:ident, $op:tt, $proxied_trait:tt, $proxied_func:ident, $scalar:ident, $scalar_ref:ident < $gen:tt >) => {
+        impl $proxied_trait<Self> for $scalar {
+            type Output = Self;
+
+            fn $proxied_func(self, rhs: Self) -> Self::Output {
+                $scalar::from(self.0.as_ref() $op rhs.0.as_ref())
+            }
+        }
+
+        impl<$gen> $proxied_trait<Self> for $scalar_ref<$gen> {
+            type Output = $scalar;
+
+            fn $proxied_func(self, rhs: Self) -> Self::Output {
+                Int256::from(self.0 $op rhs.0)
+            }
+        }
+
+        impl $trait for $scalar {
+            fn $func(&self, other: &Self) -> Option<Self> {
+                self.0.$func(*other.0).map(Into::into)
+            }
+        }
     };
 }
 
-impl_common_for_num256!(Int256, Int256Ref<'a>, I256, Int256);
+impl_checked_op!(CheckedAdd, checked_add, +, Add, add, Int256, Int256Ref<'a>);
+impl_checked_op!(CheckedSub, checked_sub, -, Sub, sub, Int256, Int256Ref<'a>);
+impl_checked_op!(CheckedMul, checked_mul, *, Mul, mul, Int256, Int256Ref<'a>);
+impl_checked_op!(CheckedDiv, checked_div, /, Div, div, Int256, Int256Ref<'a>);
+impl_checked_op!(CheckedRem, checked_rem, %, Rem, rem, Int256, Int256Ref<'a>);
+
+impl Neg for Int256 {
+    type Output = Int256;
+
+    fn neg(self) -> Self::Output {
+        Int256::from(self.0.neg())
+    }
+}
+
+impl CheckedNeg for Int256 {
+    fn checked_neg(&self) -> Option<Self> {
+        self.0.checked_neg().map(Into::into)
+    }
+}
+
+impl Zero for Int256 {
+    fn zero() -> Self {
+        Int256::from(i256::ZERO)
+    }
+
+    fn is_zero(&self) -> bool {
+        !self.0.is_negative() && !self.0.is_positive()
+    }
+}
+
+impl One for Int256 {
+    fn one() -> Self {
+        Self::from(i256::ONE)
+    }
+}
+
+impl Num for Int256 {
+    type FromStrRadixErr = ParseIntError;
+
+    fn from_str_radix(str: &str, radix: u32) -> Result<Self, Self::FromStrRadixErr> {
+        i256::from_str_radix(str, radix).map(Into::into)
+    }
+}
+
+impl Signed for Int256 {
+    fn abs(&self) -> Self {
+        self.0.abs().into()
+    }
```

Comment:
Unfortunately `num-traits` does not have `CheckedAbs`. We can expose `checked_abs` from `ethnum` directly.